### PR TITLE
Implement Course Nicknames

### PR DESCRIPTION
**Pull request by iWatger**
* added functionality for adding course nickname

Screenshot 2024-12-11 at 9 49 22 AM
Screenshot 2024-12-11 at 9 50 01 AM

### CODE REVIEW

**Comment by iWatger** [comment]
Ahh yeah I think that's pretty reasonable. I can remove that section of code for now, and stall that functionality until we have a clearer picture of how we want to implement that

**Comment by iWatger** [comment]
Removed code related to CanvasRequest and CanvasService. New screenshots!
Screenshot 2024-12-11 at 3 07 36 PM
Screenshot 2024-12-11 at 3 07 43 PM
Screenshot 2024-12-11 at 3 08 14 PM
Screenshot 2024-12-11 at 3 08 20 PM

File: CanvasPlusPlayground/Schema/Persistable/Course.swift [hunk 1]
```diff
@@ -191,7 +191,8 @@ final class Course: Cacheable {
     // We cannot use `Color` directly because it needs to conform to `PersistentModel`
     var rgbColors: RGBColors?
     var isFavorite: Bool?
-
+    var nickname: String?
+
```
Comment (rahulon12) [inline CanvasPlusPlayground/Schema/Persistable/Course.swift:195]: I'd just add a property here so you don't have to keep using the nil coalescing (`??`) operator:

```swift
var displayName: String {
    nickname ?? name ?? "Unknown Name"
}
```

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift [hunk 2]
```diff
@@ -216,6 +219,18 @@ private struct CourseListCell: View {
...
+                withAnimation {
+                    showRenameTextField = true
+                }
+            }
+            
+            Button("Remove Course Nickname") {
+                withAnimation {
+                    course.nickname = nil
+                }
+            }
```
Comment (rahulon12) [inline CanvasPlusPlayground/Features/Courses/CourseListView.swift:233]: I think a better UX might be to just have a "rename" button, and if the textfield text is empty in the alert, we set the nickname to nil. In this implementation, it seems like I'm able to have an empty nickname.

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift [hunk 3]
```diff
@@ -224,13 +239,38 @@ private struct CourseListCell: View {
...
+                    }
+                    .onDisappear {
+                        course.nickname = renameCourseFieldText
+                        renameCourseFieldText = ""
+                    }
+            
+            .padding(15)
+            
+        }
+
```
Comment (rahulon12) [inline CanvasPlusPlayground/Features/Courses/CourseListView.swift:256]: hmm idk how it works when you have a binding attached to two presentations (`alert` and `popover`). It think we can just use the `alert`, on both macOS and iOS.

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift [hunk 4]
```diff
@@ -224,13 +239,38 @@ private struct CourseListCell: View {
...
         .colorPickerSheet(
             isPresented: $showColorPicker,
             selection: $resolvedCourseColor
         ) {
             course.rgbColors = .init(color: resolvedCourseColor)
         }
+        .alert("Rename Course?", isPresented: $showRenameTextField) {
+                TextField("New name", text: $renameCourseFieldText)
+                Button("OK") {
+                    showRenameTextField = false
```
Comment (rahulon12) [inline CanvasPlusPlayground/Features/Courses/CourseListView.swift:267]:
```suggestion
                   
```
You don't need this, the alert will automatically dismiss

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift [hunk 5]
```diff
@@ -216,6 +219,18 @@ private struct CourseListCell: View {
                     course.isFavorite = !wrappedCourseIsFavorite
                 }
             }
+            
+            Button("Add Course Nickname", systemImage: "character.cursor.ibeam") {
+                withAnimation {
+                    showRenameTextField = true
```
Comment (rahulon12) [inline CanvasPlusPlayground/Features/Courses/CourseListView.swift:225]: When you present the alert, we can fill the textfield with the current nickname, if it exists.

**Review by rahulon12 — CHANGES_REQUESTED** [review]
A few small UX related changes, nice functionality though!

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift (hunk 3 above)
Comment (iWatger) [reply CanvasPlusPlayground/Features/Courses/CourseListView.swift:256]: This is it as an alert on macOS. Not too bad imo

Screenshot 2024-12-11 at 4 50 53 PM

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift (hunk 5 above)
Comment (iWatger) [reply CanvasPlusPlayground/Features/Courses/CourseListView.swift:225]: Are you meaning the textfield value, or the greyed out placeholder text? This is it with the latter

Screenshot 2024-12-11 at 4 53 22 PM

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift (hunk 2 above)
Comment (iWatger) [reply CanvasPlusPlayground/Features/Courses/CourseListView.swift]: addressed!

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift (hunk 3 above)
Comment (rahulon12) [reply CanvasPlusPlayground/Features/Courses/CourseListView.swift]: I like the alert! Looks good.

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift (hunk 5 above)
Comment (rahulon12) [reply CanvasPlusPlayground/Features/Courses/CourseListView.swift:225]: The placeholder should be the original name, yes. But if you already have a nickname the textfield should be pre-populated with that nickname. So:

```swift
renameCourseFieldText = course.nickname ?? ""
```

File: CanvasPlusPlayground/Features/Courses/CourseListView.swift [hunk 6]
```diff
@@ -216,6 +219,30 @@ private struct CourseListCell: View {
                     course.isFavorite = !wrappedCourseIsFavorite
                 }
             }
+            
+            Button("Rename \(course.name ?? "")...", systemImage: "character.cursor.ibeam") {
+                withAnimation {
```
Comment (rahulon12) [inline CanvasPlusPlayground/Features/Courses/CourseListView.swift:224]: I don't we need `withAnimation` here.

**Review by rahulon12 — APPROVED** [review]
Small changes, good to merge once those are in! This is so nice :)

**Comment by azooz2003-bit** [comment]
Love this!

**Pull request merged**